nflow-engine/src/main/java/io/nflow/engine/internal/dao/ExecutorDao.java

```diff
@@ -164,27 +163,44 @@ private void updateWithPreparedStatement(String sql) {
   }
 
   public List<WorkflowExecutor> getExecutors() {
-    return jdbc.query("select * from nflow_executor where executor_group = ? order by id asc", new RowMapper<WorkflowExecutor>() {
-      @Override
-      public WorkflowExecutor mapRow(ResultSet rs, int rowNum) throws SQLException {
-        int id = rs.getInt("id");
-        String host = rs.getString("host");
-        int pid = rs.getInt("pid");
-        DateTime started = sqlVariants.getDateTime(rs, "started");
-        DateTime active = sqlVariants.getDateTime(rs, "active");
-        DateTime expires = sqlVariants.getDateTime(rs, "expires");
-        DateTime stopped = sqlVariants.getDateTime(rs, "stopped");
-        return new WorkflowExecutor(id, host, pid, executorGroup, started, active, expires, stopped);
-      }
-    }, executorGroup);
+    return jdbc.query("select id, host, pid, started, active, expires, stopped, recovered from nflow_executor where "
+        + getExecutorGroupCondition() + " order by id asc", (rs, rowNum) -> {
+          int id = rs.getInt("id");
+          String host = rs.getString("host");
+          int pid = rs.getInt("pid");
+          var started = sqlVariants.getDateTime(rs, "started");
+          var active = sqlVariants.getDateTime(rs, "active");
+          var expires = sqlVariants.getDateTime(rs, "expires");
+          var stopped = sqlVariants.getDateTime(rs, "stopped");
+          var recovered = sqlVariants.getDateTime(rs, "recovered");
+          return new WorkflowExecutor(id, host, pid, executorGroup, started, active, expires, stopped, recovered);
+        });
   }
 
   public void markShutdown() {
     try {
-      jdbc.update("update nflow_executor " + "set expires=current_timestamp, stopped=current_timestamp "
-          + "where executor_group = ? and id = ?", executorGroup, getExecutorId());
+      jdbc.update("update nflow_executor set expires=current_timestamp, stopped=current_timestamp, recovered=null where "
+          + getExecutorGroupCondition() + " and id = ?", getExecutorId());
+    } catch (DataAccessException e) {
+      logger.warn("Failed to mark executor as stopped", e);
+    }
+  }
+
+  public Collection<Integer> getRecoverableExecutorIds() {
+    return jdbc.query(
+        "select id from nflow_executor where " + getExecutorGroupCondition() + " and id <> " + getExecutorId()
+            + " and recovered is null and " + sqlVariants.dateLtEqDiff("expires", "current_timestamp"),
+        (rs, rowNum) -> rs.getInt(1));
+  }
+
+  public void markRecovered(int recoveredExecutorId) {
+    try {
+      jdbc.update(
+          "update nflow_executor set recovered=current_timestamp where recovered is null and "
+              + sqlVariants.dateLtEqDiff("expires", "current_timestamp") + " and " + getExecutorGroupCondition() + " and id = ?",
```

Comment:
Do we need the expires or the executor group conditions here?

Reply:
If we assume that a node that has expired will never come back alive and start updating it's heartbeat, then they are not needed. Can we assume that?

Reply:
ok